src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,58 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length > 0)
+                {
+                    return length - count > 0 ? partition.Skip(length - count) : partition;
+                }
+                else if (length == 0)
```

Comment:
It looks like we don't have any tests exercising the length == -1 case.  It'd be good to add them to full exercise the new code.

Reply:
If the length == -1 it would process further with TakeLastRegularIterator<TSource>(source, count), should it behave different way in that case?

Reply:
> If the length == -1 it would process further with TakeLastRegularIterator(source, count), should it behave different way in that case?

It shouldn't behave differently, but we have no tests validating that.  We should ensure that any new such code added is fully line and branch covered by tests.  Can you please add one or more that would ensure it's appropriately covered?

You can get a coverage report by cd'ing into the test directory and doing:
```
dotnet msbuild /t:rebuildandtest /p:Coverage=true
```
That will the tests and produce a html coverage report you can then browse through to see line and branch coverage.  If you have difficulties doing so, please let me know and we can help.

Reply:
I've added one additonal IdentityTransform which seems to do the job and cover the non cheap GetCount().